### DISCUSSION TITLE
What does determine width of output layer?

### DISCUSSION THREAD

**Discussion by aoguedao** [lululxvi/deepxde#737] [Q&A]
Hi again! 

I know this question could seem a little bit naïve because for a simple example as [this from the demos](https://deepxde.readthedocs.io/en/latest/demos/pinn_forward/ode.system.html) is clear it is a 2-dimensional system of equations then net has two elements in the output layer. 

On the other hand, [inverse Poisson equation demo](https://deepxde.readthedocs.io/en/latest/demos/pinn_inverse/elliptic.inverse.field.html) is an unidimensional PDE and, even there is just one Hessian in the function `pde` and this one just returns a single element. however, the output layer also have two elements. I know this is for learning $u(x)$ and $q(x)$, but I don't understand how was defined in the dde.model that it need to learn two things ($u$ and $q$). Is it related to the dimension of the observed data? 

Thanks in advance for your help.

**Comment by forxltk** [comment]
The `pde` function returns the pde loss, but not the outputs of the NN. 

In the inverse Poisson case, `u(x)` and `q(x)` are unknown that we need to predict both of them. So there are two outputs of the NN, one is `u(x)`, another is `q(x)`. But `u(x)` and `q(x)` only follow one governing equation `ddu=q(x)`, which is the return of `pde` function.

**Comment by aoguedao** [comment]
Hi @forxltk! Thanks for your answer. I do understand the problem statement but I haven't been able to understand how (programmatically) DeepXDE knows what to learn. Documentation says _"The second argument is the network output, i.e., the solution u, q"_ but function `pde` doesn't declare the size of `y`, it just unpacking it in the first line. 

My apologies if it is a really dumb question.

**Reply by forxltk** [reply]
OK, the size of y is defined in the NN configuration. 
`net = dde.nn.PFNN([1, [20, 20], [20, 20], [20, 20], 2], "tanh", "Glorot uniform")`
Here `1` is the size of `x`, [20, 20] is the size of hidden layers, and 2 is the size of `y`.

**Reply by aoguedao** [reply]
Thank you! Just for clarifying, let's suppose I want to add another unknown function. Should I modify the function `pde` unpacking as `u, q, p = y[:, 0:1], y[:, 1:2], y[:, 1:2]` and then just change the network architecture?? For example `net = dde.nn.PFNN([1, [20, 20, 20], [20, 20, 20], [20, 20, 20], 3], "tanh", "Glorot uniform")`, right?

**Reply by forxltk** [reply]
Yes
